org.lflang/src/org/lflang/federated/FedLauncher.xtend

```diff
@@ -0,0 +1,394 @@
+        this.errorReporter = errorReporter;
+    }
+    
+    /**
+     * Return the compile command for a federate.
+     * 
+     * @param federate The federate to compile.
+     */
+    protected def String compileCommandForFederate(FederateInstance federate) {
+        throw new UnsupportedOperationException("Don't know how to compile the federates.");
+    }
+    
+    /**
+     * Return the command that will execute a federate, assuming that the current
+     * directory is the top-level project folder. This is used to create a launcher script
+     * for federates.
+     * 
+     * @param federate The federate to execute.
+     */
+    protected def String executeCommandForFederate(FederateInstance federate) {
+        throw new UnsupportedOperationException("Don't know how to execute the federates.");
+    }
+    
+    /**
+     * Create the launcher shell scripts. This will create one or two files
+     * in the output path (bin directory). The first has name equal to
+     * the filename of the source file without the ".lf" extension.
+     * This will be a shell script that launches the
+     * RTI and the federates.  If, in addition, either the RTI or any
+     * federate is mapped to a particular machine (anything other than
+     * the default "localhost" or "0.0.0.0"), then this will generate
+     * a shell script in the bin directory with name filename_distribute.sh
+     * that copies the relevant source files to the remote host and compiles
+     * them so that they are ready to execute using the launcher.
+     * 
+     * A precondition for this to work is that the user invoking this
+     * code generator can log into the remote host without supplying
+     * a password. Specifically, you have to have installed your
+     * public key (typically found in ~/.ssh/id_rsa.pub) in
+     * ~/.ssh/authorized_keys on the remote host. In addition, the
+     * remote host must be running an ssh service.
+     * On an Arch Linux system using systemd, for example, this means
+     * running:
+     * 
+     *     sudo systemctl <start|enable> ssh.service
+     * 
+     * Enable means to always start the service at startup, whereas
+     * start means to just start it this once.
+     * 
+     * On MacOS, open System Preferences from the Apple menu and 
+     * click on the "Sharing" preference panel. Select the checkbox
+     * next to "Remote Login" to enable it.
+     * 
+     * In addition, every host must have OpenSSL installed, with at least
+     * version 1.1.1a.  You can check the version with
+     * 
+     *     openssl version
+     * 
+     * @param coreFiles The files from the core directory that must be
+     *  copied to the remote machines.
+     * @param federates A list of federate instances in the federation
+     * @param federationRTIProperties Contains relevant properties of the RTI.
+     *  Can have values for 'host', 'dir', and 'user' 
+     * 
+     */
+    def createLauncher(
```

Comment:
I would suggest to break this up into multiple functions.

Reply:
This is a great suggestion. I will do that as part of the federated Python developement. I partly factored out the federated launcher here to facilitate a `FedPythonLauncher` in a modular way. Breaking this function up will make this even more modular.

Reply:
OK, sounds good to me.